### ISSUE TITLE
Phase 17.3 — PredictiveEngine: short-horizon state forecasting with uncertainty quantification

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#440]
## Phase 17.3 — `PredictiveEngine`

> **Part of Phase 17: Temporal Reasoning & Predictive Cognition**
> Upstream: Phase 17.2 `EventSequencer` (#437) · Phase 17.1 `TemporalGraph` (#434)

### Summary

`PredictiveEngine` is a short-horizon state predictor that learns temporal patterns from the `EventSequencer`'s windowed aggregates and forecasts future cognitive states with **uncertainty quantification**. It drives pre-emptive goal generation in `GoalDecomposer` and allows `CognitiveCycle` to prepare resource allocation before load spikes arrive.

---

## Enum: `PredictionStrategy`

```python
from enum import Enum, auto

class PredictionStrategy(Enum):
    LAST_VALUE         = auto()  # carry forward most-recent aggregate value
    EXPONENTIAL_SMOOTH = auto()  # EMA prediction
    LINEAR_TREND       = auto()  # 2-point linear regression extrapolation
    ENSEMBLE           = auto()  # weighted combination of all three
```

| Strategy | Complexity | Best-for |
|---|---|---|
| `LAST_VALUE` | O(1) | stable / slowly-changing signals |
| `EXPONENTIAL_SMOOTH` | O(1) | noisy signals with short memory |
| `LINEAR_TREND` | O(n) | signals with clear directional drift |
| `ENSEMBLE` | O(n) | mixed regimes; adapts weights online |

---

### `Prediction`

```python
from dataclasses import dataclass
from typing import Mapping, Any

@dataclass(frozen=True)
class Prediction:
    prediction_id: str            # uuid4 hex
    target_module: str            # e.g. "working_memory"
    predicted_state: Mapping[str, Any]  # key metric forecasts
    confidence: float             # [0.0, 1.0]
    horizon_ns: int               # how far ahead (nanoseconds)
    basis_window_ids: frozenset[str]    # WindowedAggregate IDs used
    created_at_ns: int            # wall-clock at creation
```

### `PredictionError`

```python
@dataclass(frozen=True)
class PredictionError:
    prediction_id: str
    actual_state: Mapping[str, Any]
    mse: float              # mean-squared error across predicted keys
    abs_err: float          # mean absolute error
    correct_direction: bool # did prediction get direction right?
```

### `PredictorConfig`

```python
@dataclass(frozen=True)
class PredictorConfig:
    horizon_ms: int = 1000            # prediction horizon in ms
    min_windows: int = 5              # minimum windows before predicting
    confidence_threshold: float = 0.6 # suppress predictions below this
    track_errors: bool = True         # record PredictionError for calibration
    max_history: int = 500            # max stored errors/windows per module
```

---

## Protocol: `PredictiveEngine`

```python
from typing import Protocol, runtime_checkable

@runtime_checkable
class PredictiveEngine(Protocol):
    async def train(self, windows: list[WindowedAggregate]) -> None:
        """Ingest new windowed aggregates to update internal model."""
        ...

    async def predict(self, module: str) -> Prediction | None:
        """Return forecast for module or None if insufficient data/confidence."""
        ...

    async def record_outcome(
        self,
        actual: Mapping[str, Any],
        prediction_id: str,
    ) -> PredictionError:
        """Compare actual state to prior prediction; update ensemble weights."""
        ...

    async def calibrate(self) -> dict[str, float]:
        """Recalculate smoothing alpha from recent error rates. Returns new alpha map."""
        ...

    async def stats(self) -> dict:
        """Return per-module prediction counts, mean MSE, current alpha values."""
        ...
```

---

## Implementation: `AdaptivePredictiveEngine`

```python
import asyncio, time, uuid, math
from collections import defaultdict, deque
from typing import Mapping, Any

class AdaptivePredictiveEngine:
    def __init__(
        self,
        config: PredictorConfig = PredictorConfig(),
        strategy: PredictionStrategy = PredictionStrategy.ENSEMBLE,
    ) -> None:
        self._config = config
        self._strategy = strategy
        self._lock = asyncio.Lock()
        self._windows: dict[str, deque[WindowedAggregate]] = defaultdict(
            lambda: deque(maxlen=config.max_history)
        )
        self._ema: dict[str, dict[str, float]] = defaultdict(dict)
        self._alpha: dict[str, float] = defaultdict(lambda: 0.3)
        self._weights: dict[str, list[float]] = defaultdict(
            lambda: [1/3, 1/3, 1/3]
        )
        self._errors: dict[str, deque[PredictionError]] = defaultdict(
            lambda: deque(maxlen=config.max_history)
        )
        self._pending: dict[str, tuple[str, Prediction]] = {}
        self._pred_count = self._above_threshold = self._train_count = self._cal_count = 0

    async def train(self, windows: list[WindowedAggregate]) -> None:
        async with self._lock:
            for w in windows:
                mod = w.module_id
                self._windows[mod].append(w)
                a = self._alpha[mod]
                for k, v in w.metrics.items():
                    if isinstance(v, (int, float)):
                        prev = self._ema[mod].get(k, float(v))
                        self._ema[mod][k] = a * float(v) + (1 - a) * prev
                self._train_count += 1
            TRAINING_WINDOWS.inc(len(windows))

    async def predict(self, module: str) -> Prediction | None:
        async with self._lock:
            wins = self._windows[module]
            if len(wins) < self._config.min_windows:
                return None
            wins_list = list(wins)
            predicted_state = self._run_strategy(module, wins_list)
            confidence = self._confidence(module, predicted_state)
            self._pred_count += 1
            PREDICTIONS_MADE.inc()
            if confidence < self._config.confidence_threshold:
                return None
            self._above_threshold += 1
            PREDICTIONS_ABOVE_THRESHOLD.inc()
            pred = Prediction(
                prediction_id=uuid.uuid4().hex,
                target_module=module,
                predicted_state=predicted_state,
                confidence=confidence,
                horizon_ns=self._config.horizon_ms * 1_000_000,
                basis_window_ids=frozenset(w.window_id for w in wins_list[-5:]),
                created_at_ns=time.time_ns(),
            )
            self._pending[pred.prediction_id] = (module, pred)
            return pred

    def _run_strategy(self, module: str, wins: list) -> dict[str, Any]:
        if self._strategy == PredictionStrategy.LAST_VALUE:
            return dict(wins[-1].metrics)
        if self._strategy == PredictionStrategy.EXPONENTIAL_SMOOTH:
            return dict(self._ema[module])
        if self._strategy == PredictionStrategy.LINEAR_TREND:
            return self._linear_trend(wins)
        # ENSEMBLE
        lv  = dict(wins[-1].metrics)
        ema = dict(self._ema[module])
        lt  = self._linear_trend(wins)
        w   = self._weights[module]
        keys = set(lv) | set(ema) | set(lt)
        return {k: sum(wi*v for wi,v in zip(w,[lv.get(k,0.),ema.get(k,0.),lt.get(k,0.)])) for k in keys}

    def _linear_trend(self, wins: list) -> dict[str, Any]:
        if len(wins) < 2:
            return dict(wins[-1].metrics)
        first, last = wins[0].metrics, wins[-1].metrics
        return {k: float(last[k]) + (float(last[k]) - float(first.get(k, last[k]))) for k in last}

    def _confidence(self, module: str, predicted: dict) -> float:
        vals = [float(v) for v in predicted.values() if isinstance(v, (int, float))]
        if not vals: return 0.0
        mean = sum(vals) / len(vals)
        if mean == 0.0: return 0.0
        std = math.sqrt(sum((v-mean)**2 for v in vals)/len(vals))
        return max(0.0, min(1.0, 1.0 - std/abs(mean)))

    async def record_outcome(self, actual: Mapping[str, Any], prediction_id: str) -> PredictionError:
        async with self._lock:
            entry = self._pending.pop(prediction_id, None)
            if entry is None:
                raise KeyError(f"Unknown prediction_id: {prediction_id}")
            module, pred = entry
            keys = list(pred.predicted_state)
            sq, ab, dm = [], [], []
            for k in keys:
                p, a = float(pred.predicted_state.get(k,0)), float(actual.get(k,0))
                sq.append((p-a)**2); ab.append(abs(p-a))
                last = float(list(self._windows[module])[-1].metrics.get(k, p))
                dm.append((a-last)*(p-last) > 0)
            mse = sum(sq)/len(sq) if sq else 0.0
            err = PredictionError(prediction_id=prediction_id, actual_state=dict(actual),
                                  mse=mse, abs_err=sum(ab)/len(ab) if ab else 0.0,
                                  correct_direction=bool(dm) and all(dm))
            if self._config.track_errors:
                self._errors[module].append(err)
            self._update_ensemble_weights(module, err)
            PREDICTION_MSE.set(mse)
            return err

    def _update_ensemble_weights(self, module: str, err: PredictionError) -> None:
        lr, w = 0.05, self._weights[module]
        grad = err.mse * lr
        worst = w.index(max(w))
        w = [w[i] - grad/3 if i==worst else w[i]+grad/6 for i in range(3)]
        total = sum(w)
        self._weights[module] = [x/total for x in w]

    async def calibrate(self) -> dict[str, float]:
        async with self._lock:
            result = {}
            for mod, errs in self._errors.items():
                if not errs: continue
                mean_mse = sum(e.mse for e in errs)/len(errs)
                new_a = max(0.1, min(0.9, 0.3 + mean_mse*0.5))
                self._alpha[mod] = new_a
                result[mod] = new_a
                self._cal_count += 1
                CALIBRATION_ADJUSTMENTS.inc()
            return result

    async def stats(self) -> dict:
        async with self._lock:
            return {
                "predictions_made": self._pred_count,
                "above_threshold": self._above_threshold,
                "training_windows": self._train_count,
                "calibrations": self._cal_count,
                "modules": list(self._windows.keys()),
                "alpha_map": dict(self._alpha),
                "ensemble_weights": dict(self._weights),
            }
```

### `NullPredictiveEngine`

```python
class NullPredictiveEngine:
    async def train(self, windows: list) -> None: pass
    async def predict(self, module: str) -> None: return None
    async def record_outcome(self, actual: Mapping, prediction_id: str) -> PredictionError:
        return PredictionError(prediction_id=prediction_id, actual_state=dict(actual),
                               mse=0.0, abs_err=0.0, correct_direction=True)
    async def calibrate(self) -> dict: return {}
    async def stats(self) -> dict: return {}
```

---

## Prometheus Metrics

| Metric | Type | Description |
|---|---|---|
| `asi_predictions_made_total` | Counter | All predictions generated (incl. below threshold) |
| `asi_predictions_above_threshold_total` | Counter | Predictions surfaced to callers |
| `asi_prediction_mse_gauge` | Gauge | Rolling MSE of most recent outcome |
| `asi_calibration_adjustments_total` | Counter | Times `calibrate()` updated an alpha value |
| `asi_training_windows_processed_total` | Counter | `WindowedAggregate` batches fed to `train()` |

---

## 12 Test Targets

| # | Test | Validates |
|---|---|---|
| 1 | `test_train_updates_ema` | EMA converges after repeated `train()` calls |
| 2 | `test_predict_returns_none_below_min_windows` | Guard on `min_windows` |
| 3 | `test_predict_returns_none_below_confidence` | Confidence threshold suppression |
| 4 | `test_last_value_strategy` | LAST_VALUE returns most-recent window metrics |
| 5 | `test_linear_trend_extrapolates` | LINEAR_TREND adds delta to last step |
| 6 | `test_ensemble_weights_update_on_error` | `_update_ensemble_weights` shifts weights after high MSE |
| 7 | `test_low_window_returns_none` | Under `min_windows` → `predict()` → None |
| 8 | `test_record_outcome_computes_mse` | MSE = mean of squared diffs across keys |
| 9 | `test_record_outcome_unknown_id_raises` | `KeyError` on unknown `prediction_id` |
| 10 | `test_calibrate_raises_alpha_on_high_error` | High MSE → alpha increases after `calibrate()` |
| 11 | `test_null_predict_always_returns_none` | `NullPredictiveEngine.predict()` → None |
| 12 | `test_stats_returns_expected_keys` | `stats()` keys match documented schema |

---

## Related
- Phase 17.1 `TemporalGraph` → #434
- Phase 17.2 `EventSequencer` → #437
- Phase 17 Planning → Discussion #433
- Wiki: `Phase-17-Predictive-Engine` (138th page)

**Comment by web3guru888** [comment]
## Implementation Notes — Phase 17.3 `PredictiveEngine`

### EMA Formula

```
ema_t = alpha * x_t + (1 - alpha) * ema_{t-1}
alpha = 2 / (window_count + 1)   # standard Wilder smoothing
```

Starting from `alpha = 0.3` (configurable via `calibrate()`). After N training calls:

```
alpha=0.3, N=10 windows → effective lookback ≈ 5.7 windows
alpha=0.1, N=10 windows → effective lookback ≈ 19 windows (slower reacting)
alpha=0.9, N=10 windows → effective lookback ≈ 1.2 windows (very reactive)
```

### Linear Trend (2-point regression)

Uses only first and last aggregates in the window deque:
```python
v_next = v_last + (v_last - v_first)  # extrapolate one delta-step forward
```
This avoids OLS complexity while capturing directional momentum. For noisy signals, prefer EMA or ENSEMBLE.

### Ensemble Weighting (online gradient descent)

Initial weights: `[1/3, 1/3, 1/3]` for `[LAST_VALUE, EMA, LINEAR_TREND]`.

On each `record_outcome()`:
```
worst = argmax(weights)
weights[worst] -= mse * lr / 3      # penalise highest-weight strategy
weights[others] += mse * lr / 6     # distribute credit to others
weights = normalise(weights)         # sum → 1.0
```

Learning rate `lr = 0.05` keeps updates stable; tunable in a future `PredictorConfig.lr` field.

### Confidence Calculation

```python
confidence = 1.0 - (std_dev / abs(mean))  # clamped to [0.0, 1.0]
```

- **High confidence** → predicted values cluster tightly (low variance relative to mean)
- **Low confidence** → spread predicted values → suppress prediction (return `None`)
- Edge case: `mean == 0.0` → always returns `0.0` (avoids division by zero)

### `calibrate()` Trigger Strategy

Call `calibrate()`:
1. **Every N predictions** (e.g. `N = max_history // 2` — at half-buffer saturation)
2. **On MSE spike** — when `record_outcome()` returns `mse > threshold`, call `calibrate()` inline

Production wiring in `CognitiveCycle._run_step()`:
```python
if self._pred_engine._pred_count % 250 == 0:
    await self._pred_engine.calibrate()
```

### `mypy` Narrowing Table

| Return Type | Context | Narrowing Pattern |
|---|---|---|
| `Prediction \| None` | `predict()` caller | `if pred := await engine.predict(m):` |
| `Mapping[str, Any]` vs `dict` | `record_outcome(actual=...)` | accept both; `dict(actual)` for frozen storage |
| `dict[str, float]` | `calibrate()` return | direct use — no narrowing needed |
| `list[float]` | `_weights[module]` | `defaultdict(lambda: [1/3,1/3,1/3])` always returns list |

### Test Skeletons

```python
async def test_ensemble_weights_update_on_error():
    engine = AdaptivePredictiveEngine(
        config=PredictorConfig(min_windows=2, confidence_threshold=0.0),
        strategy=PredictionStrategy.ENSEMBLE,
    )
    # seed 3 windows with rising values
    wins = [make_window("m", {"load": float(i)}) for i in range(3)]
    await engine.train(wins)
    pred = await engine.predict("m")
    assert pred is not None
    # inject a large error to shift ensemble weights
    await engine.record_outcome({"load": 999.0}, pred.prediction_id)
    w = engine._weights["m"]
    # worst strategy weight should have decreased
    assert max(w) < 1/3 + 0.05  # no single strategy dominates

async def test_low_window_returns_none():
    engine = AdaptivePredictiveEngine(
        config=PredictorConfig(min_windows=5, confidence_threshold=0.0)
    )
    await engine.train([make_window("m", {"x": 1.0})])  # only 1 window
    result = await engine.predict("m")
    assert result is None  # below min_windows guard
```

### Phase 17 Sub-phase Tracker

| Phase | Component | Issue | Status |
|---|---|---|---|
| 17.1 | TemporalGraph | #434 | 🟡 open |
| 17.2 | EventSequencer | #437 | 🟡 open |
| 17.3 | PredictiveEngine | #440 | 🟡 open |
| 17.4 | ForecastScheduler | TBD | 🔲 |
| 17.5 | TemporalOrchestrator | TBD | 🔲 |

**Comment by web3guru888** [comment]
### Phase 17.4 Filed — PredictiveEngine → SchedulerCortex Integration

Phase 17.4 `SchedulerCortex` is now tracked in issue [#443](https://github.com/web3guru888/asi-build/issues/443).

#### PredictiveEngine → SchedulerCortex Integration Pattern

The key integration point is `SchedulerCortex.prediction_loop()`, which consumes `PredictiveEngine.predict()` outputs:

```python
# SchedulerCortex.prediction_loop() — background coroutine
async def prediction_loop(self, module: str, action: str) -> None:
    while True:
        prediction = self._pe.predict(module)       # ← calls PredictiveEngine
        if prediction is not None and prediction.confidence > 0.8:
            future_ns = time.time_ns() + int(prediction.predicted_value * 1e9)
            task = ScheduledTask(
                task_id=f"pre_{module}_{future_ns}",
                module=module,
                action=action,
                priority=TaskPriority.NORMAL,
                earliest_start_ns=future_ns,
                deadline_ns=future_ns + 5_000_000_000,  # 5 s grace window
                payload={},
                prediction_basis=str(id(prediction)),    # traceability link
            )
            await self.submit(task)
        await asyncio.sleep(self._cfg.tick_ms / 1000)
```

**Data contract**:
- `PredictiveEngine.predict(module)` → `Prediction | None`
- `Prediction.confidence` threshold: **0.8** (configurable in SchedulerConfig)
- `Prediction.predicted_value`: seconds until next activation → converted to nanoseconds offset
- `prediction_basis` field carries traceability back to the specific prediction for audit purposes

#### Phase 17 Sub-phase Tracker

| Sub-phase | Component | Issue | Status |
|---|---|---|---|
| 17.1 | TemporalGraph | #434 | 🟡 In Progress |
| 17.2 | EventSequencer | #437 | 🟡 In Progress |
| **17.3** | **PredictiveEngine** | **#440** | 🟡 **In Progress** |
| 17.4 | SchedulerCortex | #443 | 🟡 In Progress |
| 17.5 | TemporalOrchestrator | — | 🔲 Pending |
